aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/serialization/AwsProxyRequestConverter.java

```diff
@@ -0,0 +1,62 @@
+package com.amazonaws.serverless.proxy.internal.serialization;
+
+import java.util.Map;
+import com.amazonaws.serverless.proxy.model.AwsProxyRequest;
+import com.amazonaws.serverless.proxy.model.Headers;
+import com.amazonaws.serverless.proxy.model.MultiValuedTreeMap;
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.type.TypeFactory;
+import com.fasterxml.jackson.databind.util.Converter;
+
+/**
+ * Populates multi-valued collections from single-valued collections if the former are not given
+ * or are empty and the latter are present and populated. This is a real-life scenario when
+ * an ALB targets a Lambda function and `lambda.multi_value_headers.enabled` is not set to `true`
+ * on the target group. 
+ */
+public class AwsProxyRequestConverter implements Converter < AwsProxyRequest, AwsProxyRequest > {
+  @Override
+  public AwsProxyRequest convert(AwsProxyRequest value) {
+        if (value == null)
+            return null;
+
+        if ((value.getMultiValueHeaders() == null || value.getMultiValueHeaders().isEmpty()) &&
+                value.getHeaders() != null && !value.getHeaders().isEmpty()) {
+            copySingleValueHeadersToMultiValueHeaders(value);
```

Comment:
Can we move this code from the converter to `AwsProxyRequest`, maybe inside the getter? I think currently the fallback logic is somewhat hidden in the converter and it's hard to see for people what really came in. Does it make sense?

Reply:
Yes, that makes sense. It's the first thing I tried! Rationale for why I changed approaches is below. I'd like to humbly suggest that we add comments that call out the converter as opposed to embedding logic in the getters. But of course, I'll defer to your judgment.

## Short Version

When I put the logic in the getters, I found that computing the values on the fly after the request object was returned by the deserializer creates unexpected interactions between object attributes that would surprise developers at best and create hard-to-find bugs at worst. I chose the converter approach because it guarantees that the request object is fully "baked" after the deserializer returns the object, and there would be no surprises to developers downstream. I agree that the code is a little less obvious, but I think it will also be more robust, so the converter is a worthy tradeoff, albeit with added comments.

## Long Version

When I put the logic in the getters, I ended up with code like this (in `AwsProxyRequest.java`):

    public MultiValuedTreeMap<String, String> getMultiValueQueryStringParameters() {
        if((multiValueQueryStringParameters==null || multiValueQueryStringParameters.isEmpty())
                && queryStringParameters != null
                && !queryStringParameters.isEmpty()) {
            multiValueQueryStringParameters = convertSingleValuesToMultiValues(queryStringParameters);
        }
        return multiValueQueryStringParameters;
    }

    public void setMultiValueQueryStringParameters(
            MultiValuedTreeMap<String, String> multiValueQueryStringParameters) {
        this.multiValueQueryStringParameters = multiValueQueryStringParameters;
    }

So the values of the multi-valued fields are not actually "baked" when the `AwsProxyRequest` object is returned from the deserializer. Rather, the value of the multi-valued fields depend on the values of their respective single-valued fields, and the multi values are computed at the time of the first call to their getters when the single-valued field is also populated. If no multi values are included in the JSON being deserialized, then these call sequences on the request object returned from deserialization end up with different values in the multi-valued fields:

    // multiValueQueryStringParameters is {"foo":"bar"} at end of code block
    apr.setQueryStringParameters(Map.of("foo", "bar"));
    apr. getMultiValueQueryStringParameters();
    apr.setQueryStringParameters(Map.of("alpha", "bravo"));

    // multiValueQueryStringParameters is {"alpha":"bravo"} at end of code block
    apr.setQueryStringParameters(Map.of("foo", "bar"));
    apr.setQueryStringParameters(Map.of("alpha", "bravo"));
    apr. getMultiValueQueryStringParameters();

I thought that violated the principle of least surprise. No one would expect assignments to one field to affect the values retrieved from another field, and rightfully so.

You could avoid this by recomputing the multi-valued fields on every call to the getter (as opposed to storing it in the instance field), like this:

    public MultiValuedTreeMap<String, String> getMultiValueQueryStringParameters() {
        if((multiValueQueryStringParameters==null || multiValueQueryStringParameters.isEmpty())
                && queryStringParameters != null
                && !queryStringParameters.isEmpty()) {
           return convertSingleValuesToMultiValues(queryStringParameters);
        }
        return multiValueQueryStringParameters;
    }

    public void setMultiValueQueryStringParameters(
            MultiValuedTreeMap<String, String> multiValueQueryStringParameters) {
        this.multiValueQueryStringParameters = multiValueQueryStringParameters;
    }

But now we're doing more (potentially a lot more) allocations and computations in getters, which isn't great either.

We could assign the multi-valued fields in the setter for the single-valued fields, but now we're potentially clobbering expected values in the multi-valued fields.

We could also potentially work around the issue by tracking when each field is assigned with a separate flag variable, but that would make the order in which getters and setters are called significant, and JSON keys are unordered, by definition.

In my opinion, the core problem is that the request object's attributes are not "baked" after the object is returned from the deserializer. What we really want is to run code to initialize the multi-valued fields one time before the deserialized object is returned to users. That is exactly what a Jackson converter does. By using the converter, we eliminate interactions between different attributes, and the request is now a POJO with no surprising behavior.

I readily acknowledge that I could be splitting hairs here that don't need splitting. If we embed the logic in the getters, then the "bug" only reveals itself in very specific circumstances which may not be likely to happen in practice. But in my opinion, for whatever that's worth, it's better to avoid the bug, even if it leads to marginally less obvious code. (I agree that we need comments, though!) But again, I'll happily defer to your judgment here.

Reply:
Wow, thank you very much for the detailed explanation. I agree but still would like to enable folks to view the real payload of the incoming event.

Just to add another option: We could also consider both single and multiValue headers in https://github.com/awslabs/aws-serverless-java-container/blob/main/aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/AwsProxyHttpServletRequest.java - e.g. modifying `getContentType` would fix the issue described in #315. How do you think about that?

Reply:
@sigpwned I'm planning to release version 1.9 next week and like to include this. Do you have feedback my comment above?

Reply:
> I'm planning to release version 1.9 next week and like to include this. Do you have feedback my comment above?

I would love to see this fix ship in the next release, too!

> Just to add another option: We could also consider both single and multiValue headers
> How do you think about that?

A quick search for `getMultiValueHeaders` in this repo flags 25 files that reference the method, and `getMultiValueQueryStringParameters` flags 4 files. The current fix -- "promoting" single-valued headers to multi-valued headers -- will work for all the existing code. If we want to force everyone to consider both multi-valued and single-valued collections in all cases, we'll have change all those files. I think that would be pretty brittle, in the long term.

> I agree but still would like to enable folks to view the real payload of the incoming event.

OK. I think there are a couple of other options we might consider:

1. We add to `AwsProxyRequest ` methods `getVerbatimMultiValueHeaders` and `getVerbatimMultiValueQueryStringParameters` that return the original values from deserialization. Even if we "rewrite" the values of `getMultiValueHeaders` and `getMultiValueQueryStringParameters`, we never rewrite those values. This fixes the bug in existing code while still exposing the original values if users need them.
2. We add to `AwsProxyRequest ` methods `getAllHeaders` and `getAllQueryStringParameters` that receive the "rewritten" values for headers, and then refactor the existing code base to use these new methods instead of the existing `getMultiValueHeaders` and `getMultiValueQueryStringParameters` methods. This will fix the bug in the library, but external users referring to these methods will need to update their code.

I think these are both good options. I prefer Option 1 (heavily commented) because it makes existing code work like users expect with no code changes, but I'll happily defer to your judgment.